src/regularizers.jl

```diff
@@ -157,22 +157,22 @@ end
 ## prox for the simplex is derived by Chen and Ye in [this paper](http://arxiv.org/pdf/1101.6081v2.pdf)
 type simplex<:Regularizer
 end
-function prox!(r::simplex,u::AbstractArray,alpha::Number)
+function prox(r::simplex,u::AbstractArray,alpha::Number)
     n = length(u)
     y = sort(u, rev=true)
     ysum = cumsum(y)
-    t = ysum[end]/n
+    t = (ysum[end]-1)/n
     for i=1:n-1
         if (ysum[i] - 1)/i >= y[i+1]
             t = (ysum[i] - 1)/i
             break
         end
     end
-    u = max(u - t, 0)
```

Comment:
the inplace `prox!` wasn't working because this line overwrites the pointer to u, rather than mutating the entries of u. it would work if you replaced this line with
```
for i=1:n
  u[i] = max(u-t, 0)
end
```